### PR TITLE
chore: Refactor token encoder/encrypter

### PR DESCRIPTION
**Pull request by craigpastro**
Refactor the token encoder.
- Add a encrypter interface and gcm and noop implementations
- Try to be consistent in `NewCommand` arguments

## Review Checklist
- [ ] I have added documentation for new/changed functionality in this PR or in [openfga.dev](https://github.com/openfga/openfga.dev)
- [ ] The correct base branch is being used, if not `main`

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -106,12 +106,12 @@ type TLSConfig struct {
 // New creates a new Server which uses the supplied backends
 // for managing data.
 func New(dependencies *Dependencies, config *Config) (*Server, error) {
-	tokenEncoder := dependencies.TokenEncoder
-	if tokenEncoder == nil {
-		tokenEncoder = encoder.NewBase64Encoder()
+	tokenEncrypter := dependencies.TokenEncrypter
+	if tokenEncrypter == nil {
```
Comment (craigpastro) [inline server/server.go:110]: By the way, why do we check these and not, say, datastore? Are these checks useful, or should we just fail fast?

File: pkg/encoder/base64.go [hunk 2]
```diff
@@ -6,16 +6,18 @@ import "encoding/base64"
...
+
 // NewBase64Encoder constructs an Encoder that implements a base64 encoding as specified
 // by the encoding/base64 package.
 func NewBase64Encoder() *Base64Encoder {
 	return &Base64Encoder{}
 }
 
-func (b *Base64Encoder) Encode(val []byte) (string, error) {
-	return base64.URLEncoding.EncodeToString(val), nil
+func (e *Base64Encoder) Decode(s string) ([]byte, error) {
```
Comment (jon-whit) [inline pkg/encoder/base64.go:17]: Can we add some godoc here such as? I'm going to start motivating us to add public godoc for anything not in the `internal` package.

```
// Decode base64 URL decodes the provided string.
```

File: pkg/encoder/base64.go [hunk 3]
```diff
@@ -6,16 +6,18 @@ import "encoding/base64"
...
+	return base64.URLEncoding.DecodeString(s)
 }
 
-func (b *Base64Encoder) Decode(s string) ([]byte, error) {
-	return base64.URLEncoding.DecodeString(s)
+func (e *Base64Encoder) Encode(data []byte) (string, error) {
```
Comment (jon-whit) [inline pkg/encoder/base64.go:21]: Same comment as above. Let's add some godoc.

```
// Encode base64 URL encodes the provided byte slice and returns the encoded value as a string.
```

File: pkg/encoder/encrypter.go [hunk 4]
```diff
@@ -1,89 +1,28 @@
...
-		return "", nil
-	}
+var _ Encrypter = (*NoopEncrypter)(nil)
 
-	encrypted, err := encrypt(e.cipherMode, data)
-	if err != nil {
-		return "", err
-	}
-	return encode(encrypted), nil
+func NewNoopEncrypter(encoder Encoder) *NoopEncrypter {
```
Comment (jon-whit) [inline pkg/encoder/encrypter.go:14]: This comments applies here and to the `gcm_encrypter`. It feels weird to me to have the encrypter structs take an `encoder`. This breaks a strong separation of concerns.

If we want to additionally encode the encrypted value returned from the encrypter, then that should be the responsibility of the caller, not this struct. I think, compositionally, we should be passing in a TokenEncrypter and internally constructing a `base64` encoder when constructing a new OpenFGA server. For example,

```
package server

func New(dependencies *Dependencies, ...) (*Server, error) {
    ...
    tokenEncrypter  := dependencies.TokenEncrypter

    s := &Server{
        tokenEncoder: encoder.NewBase64Encoder()
        tokenEncrypter: tokenEncrypter,
        ... // other fields
    }

    return &s, nil
}
```

File: pkg/encoder/base64.go (hunk 2 above)
Comment (craigpastro) [reply pkg/encoder/base64.go:17]: Good point. Added! Thanks.

File: pkg/encoder/base64.go (hunk 3 above)
Comment (craigpastro) [reply pkg/encoder/base64.go:21]: Good point. Added! Thanks.

File: pkg/encoder/encrypter.go [hunk 5]
```diff
@@ -1,89 +1,28 @@
 package encoder
```
[the remark this answers is not in the document]
Comment (craigpastro) [reply pkg/encoder/encrypter.go:1]: Thanks! Will move as part of the refactoring. Thank you!

File: pkg/encoder/encrypter.go (hunk 4 above)
Comment (jon-whit) [reply pkg/encoder/encrypter.go:14]: Sweet, no problem! Let me know when you've done the refactor and I'll finalize my review 👍 . Thanks!

Comment (craigpastro) [reply pkg/encoder/encrypter.go]: Refactored. Thanks again. It's much better now I think. Please check it out when you have time.

File: pkg/encoder/token_encoder.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+type TokenEncoder struct {
+	encrypter encrypter.Encrypter
+	encoder   Encoder
+}
+
+var _ Encoder = (*TokenEncoder)(nil)
+
+// NewTokenEncoder constructs an Encoder that includes an encrypter and encoder.
```
Comment (jon-whit) [inline pkg/encoder/token_encoder.go:14]: Did you mean?

// NewTokenEncoder constructs a TokenEncoder with the provided encrypter and encoder.

Comment (craigpastro) [reply pkg/encoder/token_encoder.go:14]: Yes, that's better. Thanks!

File: server/commands/read_authzmodels.go [hunk 7]
```diff
@@ -13,15 +13,15 @@ import (
 
 type ReadAuthorizationModelsQuery struct {
 	backend storage.AuthorizationModelReadBackend
-	encoder encoder.Encoder
 	logger  logger.Logger
+	encoder encoder.Encoder
 }
 
-func NewReadAuthorizationModelsQuery(backend storage.AuthorizationModelReadBackend, encoder encoder.Encoder, logger logger.Logger) *ReadAuthorizationModelsQuery {
+func NewReadAuthorizationModelsQuery(backend storage.AuthorizationModelReadBackend, logger logger.Logger, encoder encoder.Encoder) *ReadAuthorizationModelsQuery {
```
Comment (elbuo8) [inline server/commands/read_authzmodels.go:20]: ?

File: server/commands/read_authzmodels.go [hunk 8]
```diff
@@ -13,15 +13,15 @@ import (
...
 	return &ReadAuthorizationModelsQuery{
 		backend: backend,
-		encoder: encoder,
 		logger:  logger,
+		encoder: encoder,
```
Comment (elbuo8) [inline server/commands/read_authzmodels.go:24]: ?

File: server/commands/read_tuples.go [hunk 9]
```diff
@@ -11,29 +11,29 @@ import (
...
 	backend storage.TupleBackend
-	encoder encoder.Encoder
 	logger  logger.Logger
+	encoder encoder.Encoder
 }
 
-// NewReadTuplesQuery creates a ReadTuplesQuery with specified `tupleBackend` to use for storage
-func NewReadTuplesQuery(backend storage.TupleBackend, encoder encoder.Encoder, logger logger.Logger) *ReadTuplesQuery {
+// NewReadTuplesQuery creates a ReadTuplesQuery with specified `tupleBackend` to use for storage.
+func NewReadTuplesQuery(backend storage.TupleBackend, logger logger.Logger, encoder encoder.Encoder) *ReadTuplesQuery {
```
[the remark this answers is not in the document]
Comment (craigpastro) [reply server/commands/read_tuples.go:22]: Nice. Thanks! Done.

**Pull request merged**